### DISCUSSION TITLE
How to Permanently Modify Layer Visibility in PDFs

### DISCUSSION THREAD

**Discussion by StefanIonutTasca** [pymupdf/PyMuPDF#2776] [Q&A]
In the example below, I am trying to manipulate the visibility of a specific layer in a PDF file. I use the screenshot_page function to generate screenshots of the page with the layer visibility adjusted, and these screenshots reflect the changes correctly. However, when I use the save_pdf_with_layer function to save a new PDF with the adjusted layer visibility, the saved PDF does not reflect these changes.

**Expected Behavior**
The saved PDFs should reflect the changes made to layer visibility, showing or hiding specific layers as adjusted in the code.

**Actual Behavior**
The screenshots generated reflect the changes to layer visibility correctly. However, the PDFs saved using the save_pdf_with_layer function do not reflect these changes; the layers do not appear as expected based on the visibility settings.

import fitz  

def list_all_ocg_ids(pdf_path):
    """
    List all the OCG IDs (Optional Content Group IDs) in the given PDF.

    Args:
    - pdf_path (str): Path to the PDF file.

    Returns:
    - List[str]: List of OCG IDs.
    """
    # Open the PDF
    doc = fitz.open(pdf_path)

    # Fetch the OCGs
    ocgs = doc.get_ocgs()

    # Extract and return the OCG IDs
    ocg_ids = list(ocgs.keys())

    # Close the document
    doc.close()
    
    return ocg_ids

def save_pdf_with_layer(pdf, output_pdf_path, ocg_id, action):
    """Save a copy of the PDF with a specific OCG state."""
    
    # Adjust the visibility of the OCG based on the action
    ocgs = pdf.get_ocgs()
    for id, ocg in ocgs.items():
        if id == ocg_id:
            visibility = True if action == 2 else False
            for page in pdf:
                page.set_layer_visibility(ocg, visibility)
    
    pdf.save(output_pdf_path)
    print(f"Saved PDF to {output_pdf_path} with OCG ID {ocg_id} set to {'ON' if action == 2 else 'OFF'}")

def screenshot_page(pdf, output_image_path, ocg_id, action, page_number=0):
    """Generate a screenshot of the specified page with a specific OCG state."""
    page = pdf.load_page(page_number)
    
    # Adjust the visibility of the OCG based on the action
    ocgs = pdf.get_ocgs()
    for id, ocg in ocgs.items():
        if id == ocg_id:
            visibility = True if action == 2 else False
            page.set_layer_visibility(ocg, visibility)
    
    pix = page.get_pixmap()
    pix.save(output_image_path)
    print(f"Saved screenshot to {output_image_path}")

list_all_ocg_ids("TEST.pdf")
doc = fitz.open("TEST.pdf")
for item in doc.layer_ui_configs():
    print(item)
print()

# Set OCG with ID 2 to ON
doc.set_layer_ui_config(1, action=2)

# Take a screenshot of the first page reflecting the change
screenshot_page(doc, "screenshot.png", 1, 0)
save_pdf_with_layer(doc, "output_ON.pdf", 2, 2)

# Set OCG with ID 2 to OFF
doc.set_layer_ui_config(1, action=0)

# Take another screenshot of the first page reflecting the change
screenshot_page(doc, "screenshot2.png", 1, 0)
save_pdf_with_layer(doc, "output_OFF.pdf", 2, 0)

doc.close()

**Comment by JorjMcKie** [comment]
`fitz.Page` has no method `set_layer_visibility()`. Where does this come from?
Don't say it was ChatGPT ... 😉.

If you use `_ui_` method(s) to modify the layer, then this is temporary only. That setting will not become part of the PDF file.
For permanent changes use [doc.set_layer()](https://pymupdf.readthedocs.io/en/latest/document.html#Document.set_layer).

**Comment by StefanIonutTasca** [comment]
I tried to use this:https://github.com/pymupdf/PyMuPDF-Utilities/blob/master/jupyter-notebooks/optional-content.ipynb,

**In the tutorial I found:**
doc.set_layer_ui_config(number, action): Use this to change the state of one of the OCGs, numbered as shown. This method does the same thing as offered by supporting PDF viewers.
action is an integer with the following meanings: 0 = set on (default), 1 = toggle on/off, 2 = set off.

I would like to stick to UI because my end-goal is to create an automation based on it.

**Reply by JorjMcKie** [reply]
If you don't permanantly set the OCG as described, the current situation will not arrive in the file.
Pleas consult the documentation [here](https://pymupdf.readthedocs.io/en/latest/document.html#Document.set_layer_ui_config).
